src/subdomains/core/buy-crypto/process/entities/buy-crypto.entity.ts

```diff
@@ -243,6 +248,36 @@ export class BuyCrypto extends IEntity {
   // --- ENTITY METHODS --- //
 
   calculateOutputReferenceAmount(price: Price): this {
+    if (
+      Config.exchangeRateFromLiquidityOrder.includes(this.outputAsset.name) &&
+      this.liquidityPipeline &&
+      ![LiquidityManagementPipelineStatus.FAILED, LiquidityManagementPipelineStatus.STOPPED].includes(
+        this.liquidityPipeline.status,
+      )
+    ) {
+      if (
+        this.liquidityPipeline.status !== LiquidityManagementPipelineStatus.COMPLETE ||
+        !this.liquidityPipeline.orders?.length
+      )
+        throw new Error('LiquidityPipeline not completed');
+
+      const pipelinePrice = this.liquidityPipeline.orders[0].exchangePrice;
+      const filteredPriceSteps = price.steps.filter(
+        (s) => s.from !== pipelinePrice.source || s.to !== pipelinePrice.target,
+      );
+      const filteredPrice = filteredPriceSteps.reduce((prev, curr) => prev * curr.price, 1);
+
+      const totalPrice = Price.join(
+        this.liquidityPipeline.orders[0].exchangePrice,
+        Price.create(this.inputReferenceAsset, pipelinePrice.source, filteredPrice),
+      );
+
+      this.outputReferenceAmount = totalPrice.convert(this.inputReferenceAmountMinusFee, 8);
+      this.priceStepsObject = [...this.inputPriceStep, ...filteredPriceSteps, ...pipelinePrice.steps];
```

Comment:
Maybe it would be simpler to use this?
```ts
const totalPriceValue = [...filteredPriceSteps, pipelinePrice].reduce((prev, curr) => prev * curr.price, 1);
const totalPrice = Price.create(this.inputReferenceAsset, this.outputAsset.name, totalPriceValue);
```